For the next exercise, you are going to use the “airline_costs.csv” dataset.
The dataset has the following attributes:\
i. Airline name\
ii. Length of flight in miles\
iii. Speed of plane in miles per hour\
iv. Daily flight time per plane in hours\
v. Customers served in 1000s\
vi. Total operating cost in cents per revenue ton-mile\
vii. Revenue in tons per aircraft mile\
viii. Ton-mile load factor\
ix. Available capacity\
x. Total assets in $100,000s\
xi. Investments and special funds in $100,000s\
xii. Adjusted assets in $100,000s\
(Implement this exercise in Python language; import ‘pandas’, ‘statsmodels.api’ libraries)
Use a linear regression model to predict the number of customers each airline serves
from its length of flight and daily flight time per plane. Next, build another regression
model to predict the total assets of an airline from the customers served by the airline.
Do you have any insight about the data from the last two regression models? 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm

In [2]:
airline_costs = pd.read_csv('C:/Users/samsc/Desktop/ADS-500B/Data_and_Code2/airline_costs.csv', sep=',')

In [3]:
airline_costs

Airline  FlightLength  PlaneSpeed  DailyFlightTime  CustomersServed  \
0    All-American            57         133             6.10            20200   
1        American           270         216             6.93            56928   
2         Bonanza           100         140             4.45              183   
3         Braniff           176         182             6.60            11869   
4         Capital           142         167             7.47            41097   
5         Central            51         134             4.67             1757   
6             C&S           175         175             8.60            18000   
7        Colonial           112         150             6.87            13500   
8     Continental           131         179             6.50             3831   
9           Delta           174         191             7.60            13119   
10        Eastern           182         187             9.50            44000   
11         Empire            59         143             4.83              451   
12       Frontier            81         141             7.47             2500   
13   Lake-Central            73         142             3.85             5405   
14  Mid-Continent           144         167             6.20             6725   
15         Mohawk            79         137             5.88             9250   
16       National           199         207             8.28            23431   
17      Northeast            94         150             6.77            16000   
18      Northwest           271         202             6.88            27000   
19       Piedmont            90         153             8.13             3362   
20        Pioneer            89         150             7.05             2050   
21       Southern            67         147             6.15             2337   
22      Southwest            54         127             5.85             5143   
23    Trans-Texas            78         147             6.08             2600   
24            TWA           293         193             6.40            51500   
25         United           270         207             6.63            48913   
26     West-Coast            58         135             4.50             1850   
27        Western           172         181             6.67             6250   
28         Inland           147         158             9.38             1000   
29        Wiggins            45         116             2.35             2500   
30      Wisconsin            69         142             5.72             6500   

    TotalOperatingCost  Revenue  LoadFactor  AvailableCapacity  TotalAssets  \
0                116.3     0.96       0.400              2.400        21.13   
1                 43.0     3.98       0.689              5.776      1436.53   
2                141.5     0.79       0.358              2.207         6.65   
3                 50.6     2.57       0.557              4.614       160.30   
4                 51.0     2.68       0.510              5.255       195.02   
5                318.5     0.35       0.167              2.096        14.02   
6                 59.2     2.17       0.558              3.889       114.16   
7                 77.0     1.68       0.505              3.327        35.34   
8                 62.3     1.70       0.537              3.166        49.74   
9                 45.3     2.74       0.598              4.582       174.32   
10                42.6     3.07       0.528              5.814      1042.58   
11               112.4     0.69       0.313              2.204         4.71   
12               125.2     0.72       0.398              1.809        17.66   
13               169.3     0.48       0.212              2.264         6.32   
14                64.8     1.53       0.565              2.708        76.12   
15               100.5     1.02       0.476              2.143        14.54   
16                42.9     3.12       0.567              5.503       181.2

In [4]:
airline_costs.columns 

Index(['Airline', 'FlightLength', 'PlaneSpeed', 'DailyFlightTime',
       'CustomersServed', 'TotalOperatingCost', 'Revenue', 'LoadFactor',
       'AvailableCapacity', 'TotalAssets', 'Investments', 'AdjustedAssets'],
      dtype='object')

1. Use a linear regression model to predict the number of customers each airline serves
from its length of flight and daily flight time per plane.

In [5]:
# y = airline_costs.iloc[:, 4] #target variables
# X = airline_costs.iloc[:, [1, 3]] #predictor variables
# X = sm.add_constant(X)
# print(type(y[1]))
# a=airline_costs.iloc[:,1]
# b=airline_costs.iloc[:,3]
# print(type(a.iloc[1]))
# print(type(b.iloc[1]))
# print(X)
y = airline_costs["CustomersServed"]
X = airline_costs[["FlightLength", "DailyFlightTime"]]

In [6]:
#create training and test data to predict accuracy of linear model

In [7]:
X_train = X.iloc[0:22]
X_test = X.iloc[22:]
X_train = sm.add_constant(X_train)
X_test = sm.add_constant(X_test)
y_train = y.iloc[0:22]
y_test = y.iloc[22:]

In [8]:
model = sm.OLS(y_train, X_train).fit()
y_pred = model.predict(X_test)
print(y_pred) #prediction
print(y_test) #actual
print(model.summary()) #not a very accurate model 

22     1933.315574
23     5990.426161
24    38645.231237
25    35712.169460
26     -334.133416
27    21221.727032
28    23249.467441
29    -6826.154341
30     3888.587243
dtype: float64
22     5143
23     2600
24    51500
25    48913
26     1850
27     6250
28     1000
29     2500
30     6500
Name: CustomersServed, dtype: int64
                            OLS Regression Results                            
Dep. Variable:        CustomersServed   R-squared:                       0.537
Model:                            OLS   Adj. R-squared:                  0.488
Method:                 Least Squares   F-statistic:                     11.02
Date:                Tue, 02 Dec 2025   Prob (F-statistic):           0.000664
Time:                        07:47:23   Log-Likelihood:                -234.60
No. Observations:                  22   AIC:                             475.2
Df Residuals:                      19   BIC:                             478.5
Df Model:                           2 

Because the model had an R-squared value of 0.537, the model did not accurately predict the number of customers served. Some of the values from y_pred had customersserved in the negatives and it is not possible to serve negative amount of customers. With that in mind there might be a few reasons as to why this model is weak. The predictors are not very strong at showing how many customers are served, which is quite possible or that there is not enough training and test data to prove that these are strong predictors for having customers on airline flights.

2. Build another regression model to predict the total assets of an airline from the customers served by the airline.

In [9]:
y=airline_costs['TotalAssets']
X=airline_costs['CustomersServed']
X_train = X.iloc[0:22]
X_test = X.iloc[22:]
X_train = sm.add_constant(X_train)
X_test = sm.add_constant(X_test)
y_train = y.iloc[0:22]
y_test = y.iloc[22:]

In [10]:
model = sm.OLS(y_train, X_train).fit()
y_pred = model.predict(X_test)
print(y_pred)
print(y_test)
print(model.summary()) #not a very accurate model 

22     -2.004946
23    -52.520410
24    918.854419
25    867.464917
26    -67.418797
27     19.985074
28    -84.303635
29    -54.506861
30     24.951203
dtype: float64
22      16.06
23      11.10
24    1217.98
25    1127.25
26      10.98
27     137.39
28      16.61
29       2.03
30      10.80
Name: TotalAssets, dtype: float64
                            OLS Regression Results                            
Dep. Variable:            TotalAssets   R-squared:                       0.729
Model:                            OLS   Adj. R-squared:                  0.716
Method:                 Least Squares   F-statistic:                     53.90
Date:                Tue, 02 Dec 2025   Prob (F-statistic):           4.28e-07
Time:                        07:47:23   Log-Likelihood:                -145.95
No. Observations:                  22   AIC:                             295.9
Df Residuals:                      20   BIC:                             298.1
Df Model:                           1   

In [11]:
#The R-squared value of this model is higher and it is about 0.73. The correlation between customers served and total assets is stronger than the 
#previous regression model. This makes sense because customers spend money on the flights which gives an airline income for assets.
#That being said doesn't neccessarily mean it is a better model. The notes of the second model suggests that there is high multicollinearity 
#which means that a lot of the factors used are redundant, which could possibly mean there is overfitting. Another strong argument that this model is 
#overfitting is that there are not a lot of data in this dataset so lower data amount means one is more prone to overfitting. Both models had 
#CustomersServed as its column, once as a predictor and the other time as a target variable. I believe that since both models did not perform great,
#the first being highly inaccurate and the second having a strong argument for overfitting, that CustomersServed might not be the best choice for figuring
#out airline costs. I think a variable using Revenue as a target variable would produce a much better model. 

In [12]:
y=airline_costs['Revenue']
X=airline_costs[['Investments', 'TotalOperatingCost', 'TotalAssets']]
X_train = X.iloc[0:22]
X_test = X.iloc[22:]
X_train = sm.add_constant(X_train)
X_test = sm.add_constant(X_test)
y_train = y.iloc[0:22]
y_test = y.iloc[22:]

In [13]:
model = sm.OLS(y_train, X_train).fit()
y_pred = model.predict(X_test)
print(y_pred)
print(y_test) 
print(model.summary()) #not a very accurate model 

22    1.450684
23    1.035904
24    4.826871
25    5.344089
26    1.245730
27    1.912331
28    1.560197
29   -4.150701
30    1.039534
dtype: float64
22    1.18
23    0.65
24    3.62
25    3.75
26    0.82
27    2.68
28    1.54
29    0.07
30    0.81
Name: Revenue, dtype: float64
                            OLS Regression Results                            
Dep. Variable:                Revenue   R-squared:                       0.841
Model:                            OLS   Adj. R-squared:                  0.814
Method:                 Least Squares   F-statistic:                     31.68
Date:                Tue, 02 Dec 2025   Prob (F-statistic):           2.15e-07
Time:                        08:31:12   Log-Likelihood:                -13.703
No. Observations:                  22   AIC:                             35.41
Df Residuals:                      18   BIC:                             39.77
Df Model:                           3                                         
Covariance

In [ ]:
#Model shows a much higher R-squared value of 0.84 and no warnings of multicollinearity